[kernel] Add RPM spec

This should track the latest stable kernel for now starting with Linux 6.3 which will be released in a couple of weeks.

https://kernel.org/

Building the Linux kernel with GitHub Actions results in `No space left on device` error.

```
+ echo BUILDING A KERNEL FOR x86_64...
+ /usr/bin/make -s 'HOSTCFLAGS=-O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection   ' 'HOSTLDFLAGS=-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes ' ARCH=x86_64 olddefconfig
USING ARCH=x86_64
BUILDING A KERNEL FOR x86_64...
+ perl -p -i -e 's/^CONFIG_BUILD_SALT.*/CONFIG_BUILD_SALT="6.3.13-200.playtron.fc38.x86_64"/' .config
+ /usr/bin/make -s 'HOSTCFLAGS=-O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection   ' 'HOSTLDFLAGS=-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes ' ARCH=x86_64 'KCFLAGS= ' WITH_GCOV=0 -j2 bzImage
vmlinux.o: warning: objtool: hyperv_init+0x42f: unreachable instruction
+ '[' 1 -eq 1 ']'
+ /usr/bin/make -s 'HOSTCFLAGS=-O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection   ' 'HOSTLDFLAGS=-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes ' ARCH=x86_64 'KCFLAGS= ' WITH_GCOV=0 -j2 modules
objcopy: drivers/net/ethernet/aquantia/atlantic/stiY4CzQ: No space left on device
btf_encoder__write_elf: failed to add .BTF section to 'drivers/net/ethernet/aquantia/atlantic/atlantic.ko': 2!
Failed to encode BTF
make[1]: *** [scripts/Makefile.modfinal:62: drivers/net/ethernet/aquantia/atlantic/atlantic.ko] Error 1
make[1]: *** Deleting file 'drivers/net/ethernet/aquantia/atlantic/atlantic.ko'
make[1]: *** Waiting for unfinished jobs....
make: *** [Makefile:1954: modules] Error 2
+ exit 1
error: Bad exit status from /var/tmp/rpm-tmp.vuVTUb (%build)
    Bad exit status from /var/tmp/rpm-tmp.vuVTUb (%build)

RPM build errors:
Error: Process completed with exit code 1.
```

GitHub Actions only provides virtual machines with 14 GB of storage space. For building a Linux kernel, you need a lot more space than that.

https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

We will have to compromise and only test building the source RPM and skip the binary RPM build.

--- a/.github/workflows/gamescope-session-upload.yml
+++ b/.github/workflows/gamescope-session-upload.yml
@@ -21,7 +21,7 @@ jobs:
         run: dnf install -y copr-cli rpm-build rpmdevtools
       - name: Create the necessary RPM build directories
         run: mkdir -p ~/rpmbuild/SOURCES/
-      - name: Copy interal source files
+      - name: Copy internal source files
         run: cp ${GITHUB_WORKSPACE}/gamescope-session/* ~/rpmbuild/SOURCES/
       - name: Download external source files
         run: spectool -g -R ${GITHUB_WORKSPACE}/gamescope-session/*.spec

--- a/.github/workflows/gamescope-session.yml
+++ b/.github/workflows/gamescope-session.yml
@@ -18,7 +18,7 @@ jobs:
         run: dnf install -y copr-cli rpm-build rpmdevtools
       - name: Create the necessary RPM build directories
         run: mkdir -p ~/rpmbuild/SOURCES/
-      - name: Copy interal source files
+      - name: Copy internal source files
         run: cp ${GITHUB_WORKSPACE}/gamescope-session/* ~/rpmbuild/SOURCES/
       - name: Download external source files
         run: spectool -g -R ${GITHUB_WORKSPACE}/gamescope-session/gamescope-session.spec

--- a/.github/workflows/kernel-upload.yml
+++ b/.github/workflows/kernel-upload.yml
@@ -1,0 +1,40 @@
+---
+name: Upload kernel source RPM
+on:
+  push:
+    branches:
+      - main
+    paths:
+      - kernel.sh
+jobs:
+  upload:
+    name: Upload kernel source RPM to Fedora Copr
+    runs-on: ubuntu-22.04
+    env:
+      FEDORA_COPR_LOGIN: ${{ secrets.FEDORA_COPR_LOGIN }}
+      FEDORA_COPR_TOKEN: ${{ secrets.FEDORA_COPR_TOKEN }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Change directory
+        run: cd $GITHUB_WORKSPACE
+      - name: Create a container
+        run: docker run -d -v $(pwd):/root --name fedora38 fedora:38 sleep infinity
+      # The Linux kernel is currently handled via a script.
+      # When we get to the point where we need to add custom patches,
+      # we will add it as an additional directory similar to the existing RPMs.
+      - name: Build the Linux kernel source RPM
+        run: docker exec fedora38 /root/kernel.sh | tee /tmp/kernel.log
+      - name: Install copr-cli
+        run: docker exec fedora38 dnf install -y copr-cli
+      - name: Create the copr-cli configuration file
+        run: |
+          echo "[copr-cli]
+          login = ${FEDORA_COPR_LOGIN}
+          username = playtron
+          token = ${FEDORA_COPR_TOKEN}
+          copr_url = https://copr.fedorainfracloud.org" | sudo tee ./copr-config
+      # The Linux kernel does not support 32-bit builds
+      # but the 64-bit build is backwards compatible with 32-bit programs.
+      # Example source RPM path: /kernel/kernel-6.3.13-200.playtron.fc38.src.rpm
+      - name: Upload the source RPM to Fedora Copr
+        run: docker exec fedora38 copr-cli --config /root/copr-config build --exclude-chroot fedora-38-i386 --exclude-chroot fedora-39-i386 gaming $(grep "Wrote:" /tmp/kernel.log  | cut -d" " -f2)

--- a/.github/workflows/kernel.yml
+++ b/.github/workflows/kernel.yml
@@ -1,0 +1,40 @@
+---
+name: Test kernel
+on:
+  push:
+    branches:
+      - !main
+    paths:
+      - kernel.sh
+jobs:
+  test:
+    name: Test kernel RPM build
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Maximize build space
+        uses: easimon/maximize-build-space@master
+        with:
+          # The root file system needs space for the container.
+          root-reserve-mb: 15120
+          temp-reserve-mb: 5120
+          # Lower the swap size to 1 GB.
+          swap-size-mb: 1024
+          remove-dotnet: 'true'
+          remove-android: 'true'
+          remove-haskell: 'true'
+          remove-codeql: 'true'
+          remove-docker-images: 'true'
+      - uses: actions/checkout@v3
+      - name: Change directory
+        run: cd $GITHUB_WORKSPACE
+      - name: Create a container
+        run: docker run -d -v $(pwd):/root --name fedora38 fedora:38 sleep infinity
+      # The Linux kernel is currently handled via a script.
+      # When we get to the point where we need to add custom patches,
+      # we will add it as an additional directory similar to the existing RPMs.
+      - name: Build the Linux kernel source RPM
+        run: docker exec fedora38 /root/kernel.sh
+      - name: Build the Linux kernel binary RPMs
+        run: docker exec fedora38 rpmbuild -bb /root/kernel/kernel.spec
+      - name: Install RPMs
+        run: docker exec fedora38 dnf -y install /root/rpmbuild/RPMS/*/*.rpm

--- a/.github/workflows/mesa-upload.yml
+++ b/.github/workflows/mesa-upload.yml
@@ -21,7 +21,7 @@ jobs:
         run: dnf install -y copr-cli rpm-build rpmdevtools
       - name: Create the necessary RPM build directories
         run: mkdir -p ~/rpmbuild/SOURCES/
-      - name: Copy interal source files
+      - name: Copy internal source files
         run: cp ${GITHUB_WORKSPACE}/mesa/* ~/rpmbuild/SOURCES/
       - name: Download external source files
         run: spectool -g -R ${GITHUB_WORKSPACE}/mesa/*.spec

--- a/.github/workflows/mesa.yml
+++ b/.github/workflows/mesa.yml
@@ -18,7 +18,7 @@ jobs:
         run: dnf install -y copr-cli rpm-build rpmdevtools
       - name: Create the necessary RPM build directories
         run: mkdir -p ~/rpmbuild/SOURCES/
-      - name: Copy interal source files
+      - name: Copy internal source files
         run: cp ${GITHUB_WORKSPACE}/mesa/* ~/rpmbuild/SOURCES/
       - name: Download external source files
         run: spectool -g -R ${GITHUB_WORKSPACE}/mesa/*.spec

--- a/.github/workflows/playtron-os-scripts-upload.yml
+++ b/.github/workflows/playtron-os-scripts-upload.yml
@@ -21,7 +21,7 @@ jobs:
         run: dnf install -y copr-cli rpm-build rpmdevtools
       - name: Create the necessary RPM build directories
         run: mkdir -p ~/rpmbuild/SOURCES/
-      - name: Copy interal source files
+      - name: Copy internal source files
         run: cp ${GITHUB_WORKSPACE}/playtron-os-scripts/* ~/rpmbuild/SOURCES/
       - name: Download external source files
         run: spectool -g -R ${GITHUB_WORKSPACE}/playtron-os-scripts/*.spec

--- a/.github/workflows/playtron-os-scripts.yml
+++ b/.github/workflows/playtron-os-scripts.yml
@@ -18,7 +18,7 @@ jobs:
         run: dnf install -y copr-cli rpm-build rpmdevtools
       - name: Create the necessary RPM build directories
         run: mkdir -p ~/rpmbuild/SOURCES/
-      - name: Copy interal source files
+      - name: Copy internal source files
         run: cp ${GITHUB_WORKSPACE}/playtron-os-scripts/* ~/rpmbuild/SOURCES/
       - name: Download external source files
         run: spectool -g -R ${GITHUB_WORKSPACE}/playtron-os-scripts/*.spec

--- a/kernel.sh
+++ b/kernel.sh
@@ -1,0 +1,29 @@
+#!/bin/bash
+
+# Install generic RPM build dependencies.
+dnf install -y fedora-packager fedpkg grubby ncurses-devel pesign rpm-build rpmdevtools 'dnf-command(builddep)'
+# Download the kernel package repository. This requires first increasing the buffer size that git uses.
+git config --global http.postBuffer 314572800
+fedpkg clone -a kernel
+cd kernel
+# This is the last commit for Linux 6.3.13 before it was rebased to Linux 6.4.4.
+git checkout e0412c4338e57bc8271c644b56a0a10f220b015c
+# Install build dependencies for the Linux kernel.
+dnf builddep -y kernel.spec
+# Download the source files.
+mkdir -p ~/rpmbuild/SOURCES/
+fedpkg sources --outdir ~/rpmbuild/SOURCES/
+cp -R -v ~/rpmbuild/SOURCES/* ./
+# Copy all local source files.
+cp -R -v ./* ~/rpmbuild/SOURCES/
+# Customize the kernel to be a Playtron release kernel.
+sed -i 's/# define buildid .local/%define buildid .playtron/g' kernel.spec
+sed -i s'/%define with_debug     %{?_without_debug:     0} %{?!_without_debug:     1}/%define with_debug 0/'g kernel.spec
+sed -i s'/%define with_debuginfo %{?_without_debuginfo: 0} %{?!_without_debuginfo: 1}/%define with_debuginfo 0/'g kernel.spec
+sed -i s'/%define with_release   %{?_with_release:      1} %{?!_with_release:      0}/%define with_release 1/'g kernel.spec
+sed -i s'/%define with_headers   %{?_without_headers:   0} %{?!_without_headers:   1}/%define with_headers 1/'g kernel.spec
+sed -i s'/with_headers 0/with_headers 1/'g kernel.spec
+# Fix permissions for files relating to RPM signing.
+/usr/libexec/pesign/pesign-authorize
+# Build a source RPM.
+fedpkg --release f38 srpm

